File: app/src/main/java/com/hellmund/primetime/utils/NotificationUtils.java
package com.hellmund.primetime.utils;

import android.annotation.TargetApi;
import android.app.AlarmManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;

import com.hellmund.primetime.R;
import com.hellmund.primetime.database.WatchlistMovie;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class NotificationUtils {

    private static final String CHANNEL_ID = "channel";
    private static final String GROUP_KEY = "group";

    @TargetApi(Build.VERSION_CODES.O)
    public static void createChannel(Context context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            return;
        }

        NotificationChannel channel = new NotificationChannel(
                "releases", "Release notifications", NotificationManager.IMPORTANCE_LOW);

        NotificationManager manager =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (manager != null) {
            manager.createNotificationChannel(channel);
        }
    }

    static Notification buildNotification(Context context, List<WatchlistMovie> movies) {
        NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle();

        for (int i = 0; i < movies.size(); i++) {
            inboxStyle.addLine(movies.get(i).getTitle());
        }

        String header = context.getResources().getQuantityString(
                R.plurals.new_releases_notif_header, movies.size(), movies.size());
        inboxStyle.setBigContentTitle(header);

        ArrayList<String> titles = new ArrayList<>();
        for (WatchlistMovie movie : movies) {
            titles.add(movie.getTitle());
        }

        StringBuilder collapsedText = new StringBuilder();
        if (!titles.isEmpty()) {
            collapsedText.append(titles.get(0));
        }

        int size = titles.size();
        for (int i = 1; i < size; i++) {
            collapsedText.append(", ").append(titles.get(i));
        }

        return new NotificationCompat.Builder(context, CHANNEL_ID)
                .setStyle(inboxStyle)
                .setContentTitle(header)
                .setContentText(collapsedText.toString())
                .setSmallIcon(R.drawable.ic_notification)
                .setColor(ContextCompat.getColor(context, R.color.app_color))
                .setGroup(GROUP_KEY)
                .build();
    }

    public static void scheduleNotifications(final Context context) {
        Intent notificationIntent = new Intent(context, NotificationPublisher.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, notificationIntent, 0);

        AlarmManager alarmMgr = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Calendar cal = DateUtils.startOfDay();
        cal.set(Calendar.HOUR_OF_DAY, 9);

        if (alarmMgr != null) {
            alarmMgr.setRepeating(AlarmManager.RTC, cal.getTimeInMillis(),
                                  AlarmManager.INTERVAL_DAY, pendingIntent);
        }
    }

    public static boolean areNotificationsEnabled(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(Constants.KEY_NOTIFICATIONS, true);
    }

}

File: app/src/main/java/com/hellmund/primetime/utils/NotificationUtils.kt
package com.hellmund.primetime.utils

import android.app.AlarmManager
import android.app.AlarmManager.INTERVAL_DAY
import android.app.AlarmManager.RTC
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build.VERSION.SDK_INT
import android.os.Build.VERSION_CODES.O
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.core.content.getSystemService
import androidx.preference.PreferenceManager
import com.hellmund.primetime.R
import com.hellmund.primetime.data.model.WatchlistMovie
import org.threeten.bp.LocalDate
import org.threeten.bp.ZoneId

object NotificationUtils {

    private const val GROUP_KEY = "releases_group"
    private const val CHANNEL_ID = "releases"

    @JvmStatic
    fun createChannel(context: Context) {
        if (SDK_INT < O) {
            return
        }

        val name = context.getString(R.string.release_notifications)
        val channel = NotificationChannel(CHANNEL_ID, name, NotificationManager.IMPORTANCE_LOW)
        context.getSystemService<NotificationManager>()?.createNotificationChannel(channel)
    }

    @JvmStatic
    fun buildNotification(context: Context, movies: List<WatchlistMovie>): Notification {
        val inboxStyle = NotificationCompat.InboxStyle()
        val titles = movies.map { it.title }
        titles.forEach { title -> inboxStyle.addLine(title) }

        val header = context.resources.getQuantityString(
            R.plurals.new_releases_notif_header, movies.size, movies.size)
        inboxStyle.setBigContentTitle(header)

        val collapsedText = titles.joinToString(", ")

        return NotificationCompat.Builder(context, CHANNEL_ID)
            .setStyle(inboxStyle)
            .setContentTitle(header)
            .setContentText(collapsedText)
            .setSmallIcon(R.drawable.ic_notification)
            .setColor(ContextCompat.getColor(context, R.color.red_400))
            .setGroup(GROUP_KEY)
            .build()
    }

    @JvmStatic
    fun scheduleNotifications(context: Context) {
        val notificationIntent = Intent(context, NotificationPublisher::class.java)
        val pendingIntent = PendingIntent.getBroadcast(context, 0, notificationIntent, 0)

        val alarmTime = LocalDate.now()
            .atStartOfDay(ZoneId.systemDefault())
            .withHour(9)
            .toInstant()

        val alarmManager = context.getSystemService<AlarmManager>()
        alarmManager?.setRepeating(RTC, alarmTime.toEpochMilli(), INTERVAL_DAY, pendingIntent)
    }

    @JvmStatic
    fun areNotificationsEnabled(context: Context): Boolean {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        return sharedPrefs.getBoolean(Preferences.KEY_NOTIFICATIONS, true)
    }

}
